src/java.base/share/classes/java/io/Console.java

```diff
@@ -592,25 +595,43 @@ public int read(char[] cbuf, int offset, int length)
 
         CHARSET = cs;
 
+        cons = instantiateConsole(istty);
+
         // Set up JavaIOAccess in SharedSecrets
         SharedSecrets.setJavaIOAccess(new JavaIOAccess() {
             public Console console() {
-                if (istty) {
-                    if (cons == null)
-                        cons = new Console();
-                    return cons;
-                }
-                return null;
-            }
-
-            public Charset charset() {
-                return CHARSET;
+                return cons;
             }
         });
     }
-    private static Console cons;
+
+    @SuppressWarnings("removal")
+    private static Console instantiateConsole(boolean istty) {
+        try {
+            // Try loading providers
+            PrivilegedAction<Console> pa = () -> {
+                var consModName = System.getProperty("jdk.console",
+                        JdkConsoleProvider.DEFAULT_PROVIDER_MODULE_NAME);
+                return ServiceLoader.load(JdkConsoleProvider.class).stream()
```

Comment:
Are we intentionally using thread context classloader (which can be different depending on which caller ends up first accessing/initializing the `java.io.Console` class) to load these services?

I initially thought that `java.io.Console` might be used/initialized early in the bootstrap process of Java so the classloader could perhaps be deterministic, but running a trivial Java application with `-verbose:class` shows that `java.io.Console` doesn't get instantiated during the launch, so that leaves this code to "first access wins" situation and maybe an "incorrect" context classloader which doesn't have access the configured `jdk.console` module may end up causing this code to default to `java.io.Console`?

```
public class Hello {
	public static void main(final String[] args) {
	}
}
```
```
java -verbose:class Hello.java
```

Instead, should we perhaps use the ModuleLayer to find this configured module and then use its classloader to load the `JdkConsoleProvider` service provider? Something like:

```
final Optional<Module> mod = ModuleLayer.boot().findModule(consModName);
// ... if not present default to java.io.Console else use the module's classloader to try and load the JdkConsoleProvider
return ServiceLoader.load(JdkConsoleProvider.class, mod.get().getClassLoader()).stream()......

Reply:
Changed it to use the boot layer `ServiceLoader.load(ModuleLayer.boot(), JdkConsoleProvider.class)`